### PR TITLE
docs(schema): add NuxtHooks interface documentation

### PR DESCRIPTION
**Pull request by Draichi**
### 🔗 Linked issue

https://github.com/nuxt/nuxt/issues/13653

### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

The section [Nuxt Hooks (build time)](https://nuxt.com/docs/api/advanced/hooks#nuxt-hooks-build-time) is not yet complete on the live docs page, and it's wrongly marked as "OK" as commented [here](https://github.com/nuxt/nuxt/issues/13653#issuecomment-1397311124)

This PR adds the documentation on the interface itself and on the lifecycle hooks docs page.

This PR is important to provide information about nuxt hooks and to avoid delving into the source code to understand the hooks like I did myself.

Please make sure the description I provided for each hook is correct

Thank you

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by danielroe** [comment]
This is great work, thank you ❤️

I'll look through and review in detail later.

File: docs/3.api/4.advanced/1.hooks.md [hunk 1]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
 
 Check the [schema source code](https://github.com/nuxt/nuxt/blob/main/packages/schema/src/types/hooks.ts#L53) for all available hooks.
 
-:NeedContribution
+Hook                     | Arguments                  | Environment     | Description
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:36]: I think we can lose the environment column as they all happen within the build server.

Later, we can refactor to perhaps list these in the normal order they are called, but I wouldn't want to block the PR for that.

File: docs/3.api/4.advanced/1.hooks.md [hunk 2]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+-------------------------|----------------------------|-----------------|-------------
+`kit:compatibility`      | `compatibility, issues`    | Server          | Allows extending compatibility checks.
+`ready`                  | `nuxt`                     | Server          | Called after nuxt initialization, when the nuxt instance is ready to work.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:39]: For consistency, let's capitalise `nuxt`, so `Nuxt`.

File: docs/3.api/4.advanced/1.hooks.md [hunk 3]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`close`                  | `nuxt`                     | Server          | Called when Nuxt instance is gracefully closing.
+`modules:before`         | -                          | Server          | Called during nuxt initialization, before install user modules.
+`modules:done`           | -                          | Server          | Called during nuxt initialization, after install user modules.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:42]:
```suggestion
`modules:before`         | -                          | Server          | Called during Nuxt initialization, before installing user modules.
`modules:done`           | -                          | Server          | Called during Nuxt initialization, after installing user modules.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 4]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`app:resolve`            | `app`                      | Server          | Called after resolve the `app` instance.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:43]:
```suggestion
`app:resolve`            | `app`                      | Server          | Called after resolving the `app` instance.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 5]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`app:templates`          | `app`                      | Server          | Called during `NuxtApp` generation, after configure user templates.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:44]:
```suggestion
`app:templates`          | `app`                      | Server          | Called during `NuxtApp` generation, to allow customizing, modifying or adding new files to the build directory (either virtually or to written to `.nuxt`).
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 6]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`app:templatesGenerated` | `app`                      | Server          | Called after templates are compiled into virtual file system (vfs).
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:45]:
```suggestion
`app:templatesGenerated` | `app`                      | Server          | Called after templates are compiled into the [virtual file system](https://nuxt.com/docs/guide/directory-structure/nuxt#virtual-file-system) (vfs).
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 7]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`build:before`           | -                          | Server          | Called before Nuxt bundle builder.
+`build:done`             | -                          | Server          | Called after Nuxt bundle builder is complete.
+`build:manifest`         | `manifest`                 | Server          | Called during the manifest build by Vite and Webpack.
+`builder:generateApp`    | `options`                  | Server          | Called before generate the app.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:49]:
```suggestion
`builder:generateApp`    | `options`                  | Server          | Called before generating the app.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 8]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`close`                  | `nuxt`                     | Server          | Called when Nuxt instance is gracefully closing.
+`modules:before`         | -                          | Server          | Called during nuxt initialization, before install user modules.
+`modules:done`           | -                          | Server          | Called during nuxt initialization, after install user modules.
+`app:resolve`            | `app`                      | Server          | Called after resolve the `app` instance.
+`app:templates`          | `app`                      | Server          | Called during `NuxtApp` generation, after configure user templates.
+`app:templatesGenerated` | `app`                      | Server          | Called after templates are compiled into virtual file system (vfs).
+`build:before`           | -                          | Server          | Called before Nuxt bundle builder.
+`build:done`             | -                          | Server          | Called after Nuxt bundle builder is complete.
+`build:manifest`         | `manifest`                 | Server          | Called during the manifest build by Vite and Webpack.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:48]:
```suggestion
`build:manifest`         | `manifest`                 | Server          | Called during the manifest build by Vite and Webpack. This allows customizing the manifest that Nitro will use to render `<script>` and `<link>` tags in the final HTML.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 9]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`builder:generateApp`    | `options`                  | Server          | Called before generate the app.
+`builder:watch`          | `event, path`              | Server          | Called on build time when option.dev is enabled.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:50]:
```suggestion
`builder:watch`          | `event, path`              | Server          | Called at build time in development when the watcher spots a change to a file or directory in the project.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 10]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`pages:extend`           | `pages`                    | Server          | Called after pages routes are resolved.
+`server:devHandler`      | `handler`                  | Server          | Called when the dev middleware is been registered on server.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:52]:
```suggestion
`server:devHandler`      | `handler`                  | Server          | Called when the dev middleware is being registered on the Nitro dev server.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 11]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`imports:sources`        | `presets`                  | Server          | Called at setup allowing modules extending sources.
+`imports:extend`         | `imports`                  | Server          | Called at setup allowing modules extending imports.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:54]:
```suggestion
`imports:sources`        | `presets`                  | Server          | Called at setup allowing modules to extend sources.
`imports:extend`         | `imports`                  | Server          | Called at setup allowing modules to extend imports.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 12]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`imports:context`        | `context`                  | Server          | Called when the Unimport context is created.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:55]:
```suggestion
`imports:context`        | `context`                  | Server          | Called when the [unimport](https://github.com/unjs/unimport) context is created.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 13]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`imports:dirs`           | `dirs`                     | Server          | Allows extending import directories.
+`components:dirs`        | `dirs`                     | Server          | Called at app:resolve allowing to extend the directories.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:57]:
```suggestion
`components:dirs`        | `dirs`                     | Server          | Called within `app:resolve` allowing to extend the directories that are scanned for auto-importable components.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 14]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`components:extend`      | `components`               | Server          | Allows extending new components.
+`nitro:config`           | `nitroConfig`              | Server          | Called at nitro init allowing to extend the configuration.
+`nitro:init`             | `nitro`                    | Server          | Called when nitro is exposed at Nuxt instance.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:60]:
```suggestion
`nitro:config`           | `nitroConfig`              | Server          | Called before initializing Nitro, allowing customization of Nitro's configuration.
`nitro:init`             | `nitro`                    | Server          | Called after Nitro is initialized, which allows registering Nitro hooks and interacting directly with Nitro.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 15]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`nitro:build:before`     | `nitro`                    | Server          | Called before build the nitro instance.
+`prerender:routes`       | `ctx`                      | Server          | Allows extending the routes to be prerender.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:62]:
```suggestion
`nitro:build:before`     | `nitro`                    | Server          | Called before building the nitro instance.
`prerender:routes`       | `ctx`                      | Server          | Allows extending the routes to be prerendered.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 16]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`build:error`            | `error`                    | Server          | Called when an error occurs at build time.
+`prepare:types`          | `options`                  | Server          | Called when nuxi is writing types.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:64]:
```suggestion
`prepare:types`          | `options`                  | Server          | Called before nuxi writes `.nuxt/tsconfig.json` and `.nuxt/nuxt.d.ts`, allowing addition of custom references and declarations in `nuxt.d.ts`, or directly modifying the options in `tsconfig.json`
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 17]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`listen`                 | `listenerServer, listener` | Server          | Called when the dev server is loading.
+`schema:extend`          | `schemas`                  | Server          | Allows to extend default schemas.
+`schema:resolved`        | `schema`                   | Server          | Allows to extend resolved schema.
+`schema:beforeWrite`     | `schema`                   | Server          | Called before write the given schema.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:68]:
```suggestion
`schema:extend`          | `schemas`                  | Server          | Allows extending default schemas.
`schema:resolved`        | `schema`                   | Server          | Allows extending resolved schema.
`schema:beforeWrite`     | `schema`                   | Server          | Called before writing the given schema.
```

File: docs/3.api/4.advanced/1.hooks.md [hunk 18]
```diff
@@ -33,7 +33,51 @@ Hook                   | Arguments           | Environment     | Description
...
+`schema:written`         | -                          | Server          | Called after the schema is written.
+`vite:extend`            | `viteBuildContext`         | Server          | Allows to extend Vite default context.
+`vite:extendConfig`      | `viteInlineConfig, env`    | Server & Client | Allows to extend Vite default config.
+`vite:serverCreated`     | `viteServer, env`          | Server & Client | Called when the vite server is created.
+`vite:compiled`          | -                          | Server & Client | Called after vite server is compiled.
+`webpack:config`         | `webpackConfigs`           | Server          | Called before configure the webpack compiler.
```
Comment (danielroe) [inline docs/3.api/4.advanced/1.hooks.md:74]:
```suggestion
`vite:serverCreated`     | `viteServer, env`          | Server & Client | Called when the Vite server is created.
`vite:compiled`          | -                          | Server & Client | Called after Vite server is compiled.
`webpack:config`         | `webpackConfigs`           | Server          | Called before configuring the webpack compiler.
```

**Review by danielroe — COMMENTED** [review]
This is phenomenal. Made a couple of small comments. If you could copy them across also to `packages/schema/src/types/hooks.ts` that would be very kind, and just check for consistent casing on webpack, Vite, Nuxt and Nitro. I think case-police (run `pnpm lint:docs`) should help with this.

(Later, we should add functionality to pull this info directly from the source so we don't have to duplicate this kind info, and so any new hook gets automatically added to docs.)

**Comment by Draichi** [comment]
@danielroe I updated the files according to your requests

**Review by danielroe — APPROVED** [review]
Thank you! ❤️

**Pull request merged**